Updated the local installation guide for Astronomer Core

jwitz opened the pull request:
This PR makes a few key changes to our existing setup guides for AC:

1. The local Quickstart now relies heavily on the Astronomer CLI
2. The production installation guide no longer mentions how to deploy DAGs (this will be in a separate doc)

All other doc changes are wording-based and shouldn't result in any change to functionality/steps from previous guides. There are still places without links and which reference "Certified" -- that will be fixed closer to actual publication. 


I'll leave questions throughout the doc for reviewers, but more generally I want to ask (especially for @ernest-kr ): What steps can we provide so that production installers can check that they successfully installed Airflow? Will they automatically be able to open `http://localhost:8080/`  because of the systemd unit file? Or are there other parts of the installation that they need to confirm using commands etc?

jwitz commented on ac/next/02_production-core.md:
```diff
@@ -0,0 +1,236 @@
+
+To make the database server accessible outside of your localhost, you may have to edit your [`/var/lib/postgres/data/pg_hba.conf`](https://www.postgresql.org/docs/10/auth-pg-hba-conf.html) file and restart Postgres. Editing this file will vary for each individual database setup. You should also understand the security implications before editing this file.
+
+If your database server is running on the same machine, you can change `peer` to `md5` to allow connections with username/password from the same machine.
```
This is unclear in our initial guide: "Same machine" as what?

paolaperaza replied:
I assume this means "on a single machine"?

jwitz commented on ac/next/02_production-core.md:
````diff
@@ -0,0 +1,236 @@
+
+```sh
+$ sudo install --owner=astro --group=astro -d /usr/local/airflow
+$ echo 'export AIRFLOW_HOME=/usr/local/airflow' | sudo tee --append ~astro/.bashrc
````
Does this command create the file structure that `astro dev init` creates? Do production installers need to recreate that structure themselves?

paolaperaza replied:
Good question.... it may just be empty?

paolaperaza commented on ac/next/01_quickstart.md:
````diff
@@ -0,0 +1,123 @@
+1. Run `export AIRFLOW_HOME=~/airflow` to give Astronomer Core a home root directory on your machine.
+
+2. Run the following command to get the latest version of the distribution on your machine:
+
+    ```sh
+    PIP_EXTRA_INDEX_URL='https://pip.astronomer.io/simple' pip3 install -U 'astronomer-certified==1.10.10.*'
+    ```
+
+    Note that if you'd like to install a specific version, you can do so by appending the version tag via the following syntax:
+
+    ```sh
+    PIP_EXTRA_INDEX_URL='https://pip.astronomer.io/simple' pip3 install -U 'astronomer-certified==1.10.10-1'
+    ```
````
Probably more room here to explain versioning - do you think that merits a separate "Astronomer Core Versioning" doc, much like we have "Upgrade Apache Airflow on Astronomer"? https://www.astronomer.io/docs/cloud/stable/customize-airflow/manage-airflow-versions

jwitz replied:
Yep, I've got an upgrade doc in the pipelines! Would definitely link here once it's completed.

paolaperaza commented on ac/next/01_quickstart.md:
````diff
@@ -0,0 +1,123 @@
+    PIP_EXTRA_INDEX_URL='https://pip.astronomer.io/simple' pip3 install -U 'astronomer-certified==1.10.10-1'
+    ```
+
+3. Run `$ airflow initdb` to build out your project directory and initialize a lightweight SQLite database for Airflow.
````
We should figure out if people could swap SQLite for another database here - or, maybe this is our default locally and we say "to use another database in production refer to x doc"

jwitz replied:
It looks like you can modify aspects of the db by running `airflow reset` after changing configurations in `airflow.cfg`, but you always need to run `airflow initdb` once to actually initialize it. For the quickstart, I think this would suffice. But if we have a doc about modifying the database in the future, this would be a good place to link that.

paolaperaza commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,123 @@
+
+## Next Steps
+
+Now that you have Astronomer Core Airflow running on your local machine, you might want to do one of the following:
```
```suggestion
Now that you have Astronomer Core running on your local machine, you might want to do one of the following:
```

paolaperaza commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,123 @@
+
+Now that you have Astronomer Core Airflow running on your local machine, you might want to do one of the following:
+
+* Set up Astronomer Core in a production environment.
+* Try out key AC and Airflow features using one of our Airflow guides.
+* Learn more about what's in the AC image.
+
+If you want all of the features of AC, plus additional metrics, security, and customer support, consider exploring our Astronomer Cloud and Enterprise offerings.
```
If we assume that anyone trying out Airflow for the first time is reading this guide, I think there's a lot of room here to expand these resources. That includes:
- Linking to Airflow Docs
- Linking to Airflow Guides (mentioned but not linked), maybe suggesting 1 or 2
- If we chose to break apart our "Production" environment doc into specific guidelines on how to do x or y action, that might influence our guidance here

paolaperaza replied:
^ Still applicable 🙂

paolaperaza commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,123 @@
+
+There are two primary ways to obtain the Astronomer Core:
+
+- [Docker Image](https://quay.io/repository/astronomer/ap-airflow?tab =tags)
```
```suggestion
- [Debian-based Docker Image](https://quay.io/repository/astronomer/ap-airflow?tab =tags)
```

paolaperaza commented on ac/next/02_production-core.md:
```diff
@@ -0,0 +1,236 @@
+
+### C. Install Astronomer Core
+
+Install the AC Python Package onto your machine by running:
```
Question to look into: What if a user already installed this locally (by following our other doc)? Do they have to do it again? Could they use a Docker image here instead?

jwitz replied:
This is something I still don't fully understand: what does this setup look like if you *want* to use Docker? Do we need a fully separate setup guide for using Docker/ Kube?

paolaperaza commented on ac/next/02_production-core.md:
````diff
@@ -0,0 +1,236 @@
+    WantedBy=multi-user.target
+    ```
+
+3. Edit `/etc/default/astronomer-certified` to contain these Environment Variables and values:
+
+   ```sh
+   AIRFLOW_HOME=/usr/local/airflow/
+   AIRFLOW__CORE__LOAD_EXAMPLES=False
+   PATH=$PATH:/home/astro/airflow-venv/bin
+   ```
````
Need to look into why we call for these default Env Vars 🤔  (`   AIRFLOW__CORE__LOAD_EXAMPLES=False`, for example, calls for all the example DAGs to not show up by default, but not sure why we're making that decision for the user)

jwitz replied:
I assume we do that because it's for production? Otherwise, we'd have to mention in a separate guide how to turn those off. We could also explain what each one does just below it and give users the discretion to modify as needed.

paolaperaza commented on ac/next/02_production-core.md:
````diff
@@ -0,0 +1,236 @@
+For Celery Executor:
+
+```
+AIRFLOW__CORE__SQL_ALCHEMY_CONN=postgresql://airflow:<your-user-password>@localhost/airflow
+AIRFLOW__CELERY__BROKER_URL=sqla+postgresql://airflow:<your-user-password>@localhost/airflow
+AIRFLOW__CELERY__RESULT_BACKEND=db+postgresql://airflow:<your-user-password>@localhost/airflow
````
Should the value for `your-user-password` here be the same?

paolaperaza commented on ac/next/02_production-core.md:
````diff
@@ -0,0 +1,236 @@
+AIRFLOW__CORE__EXECUTOR=CeleryExecutor
+```
+
+The password you specify here should be the same one you specified when prompted by the `createuser` command in step 1. If your password contains `%`, `/`, or `@` then you will need to url-escape; replace `%` with `%25`, `/` with `%2F`, and `@` with `%40`.
````
This should be cleaned up a bit

paolaperaza commented on ac/next/02_production-core.md:
````diff
@@ -0,0 +1,236 @@
+
+The password you specify here should be the same one you specified when prompted by the `createuser` command in step 1. If your password contains `%`, `/`, or `@` then you will need to url-escape; replace `%` with `%25`, `/` with `%2F`, and `@` with `%40`.
+
+#### Alternative setup options
+
+* Your Airflow user password is stored written in `/etc/default/astronomer-certified` (owned by `root:root` and `0600` permissions) on your nodes. If you'd rather use an existing credential store, such as HashiCorp's Vault, you can instead specify a command that will be run (once, at service start up) to obtain the connection string. For example:
+
+    ```
+    AIRFLOW__CORE__SQL_ALCHEMY_CONN_CMD=vault kv get -field=dsn secret/airflow-db
+    ```
+
+    For more information on this feature, read [Integrating Airflow and Hashicorp Vault](/guides/airflow-and-hashicorp-vault).
+
+* In this example setup, the Environment Variables for the Celery Executor specify using the main database for communication. Alternatively, you could use [Redis](https://redis.io/) to take some of the load off of your main database.
````
This _may_ belong as its own separate doc?

jwitz replied:
Making a note of this in clubhouse

paolaperaza reviewed (commented):
@jwitz Thanks for putting this together - it's definitely a good start. We'll chat about this in person, but we absolutely do need to document our target docs tree structure - I have some ideas after reading through this PR. In parallel to us figuring that out, though, I'd say let's also work on polishing up the local quickstart doc here before we get anyone else to review. That might include:

- Tightening up spelling, capitalization, formatting
- Clean up front-matter description
- Create a draft for a new "Install the Astronomer CLI" doc that's short/simple, and abstracts Astronomer product stuff away (and links out here)
- Try to clean up language about how we refer to Astronomer Core - i.e. _not_ `Astronomer Core Airflow`, not Certified, etc.
- Link out to Airflow docs/the project where you think it may be appropriate
- Clean up resources section at the end

While we get this polished up, we can in tandem identify content gaps and keep adding as we go. WDYT?

jwitz commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,123 @@
+
+- [Python 3](https://www.python.org/downloads/)
+- [Docker](https://www.docker.com/products/docker-desktop)
+- [Astronomer CLI](https://www.astronomer.io/docs/enterprise/v0.23/get-started/quickstart)
```
This is a can-of-worms question at the moment, as we'll have to link out regardless if we're hosting the CLI reference guide on product docs. I think our end goal would be to host all CLI docs on AC, but for the time being (ie before we have our own landing page) I don't think it hurts to link out.

jwitz commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,123 @@
+
+## Prerequisites
+
+To run Astronomer Certified locally, you'll need the following on your machine:
```
Assuming it's cool we do this when the packages/ images still reference certified?

jwitz commented:
@paolaperaza Per our discussion today, updated the local quickstart to include installation steps for the Astro CLI. Also updated the title and description of the production installation guide.

paolaperaza commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,166 @@
+
+To get started with AC on your local machine, follow this guide. We'll use the Astronomer CLI to create a local Airflow project and provision all essential components to get it running.
+
+There are two primary ways to obtain the Astronomer Core:
+
+- [Debian-based Docker Image](https://quay.io/repository/astronomer/ap-airflow?tab=tags)
+- [Python Wheel](https://pip.astronomer.io/simple/apache-airflow/)
+
+This quickstart focuses on setting up Astronomer Core via Docker, which is the what we recommend for local installations. For the Python Wheel setup, read to [Install via Python Wheel](/docs/ac/next/01_quickstart#install-via-python-wheel).
+
+For Astronomer's full collection of Docker Images, reference our public [Quay.io repository](https://quay.io/repository/astronomer/ap-airflow?tab=tags).
```
```suggestion
For every major and minor version of Apache Airflow, our team produces two Astronomer Core artifacts:

- [A Debian-based Docker image](https://quay.io/repository/astronomer/ap-airflow?tab=tags)
- [A Python Wheel](https://pip.astronomer.io/simple/apache-airflow/)

This quickstart focuses on the Astronomer Core Docker image, which is the recommended and most popular way to deploy the offering. For instructions on how to install the Python Wheel, read [Install via Python Wheel](/docs/ac/next/01_quickstart#install-via-python-wheel).

For Astronomer's full collection of Docker images, reference either of our public Docker registries:

- [Quay.io](https://quay.io/repository/astronomer/docker-airflow?tab=tags).
- [Docker Hub](https://hub.docker.com/r/astronomerio/docker-airflow)
```

paolaperaza commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,166 @@
+
+For Astronomer's full collection of Docker Images, reference our public [Quay.io repository](https://quay.io/repository/astronomer/ap-airflow?tab=tags).
+
+> **Note: Astronomer Core is also hosted on [Docker Hub](https://hub.docker.com/r/astronomerinc/ap-airflow), but pulling your image from there might result in rate limiting errors. Because of that, we recommend pulling from the Quay.io repository.
```
```suggestion
> **Note**: While you're free to pull the Astronomer Core Docker image from Docker Hub, you may encounter a [Docker pull rate limit error](https://forum.astronomer.io/t/docker-hub-rate-limit-error-toomanyrequests-you-have-reached-your-pull-rate-limit/887). To avoid limitations, we recommend referencing and pulling from Quay.io.
```

paolaperaza commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,166 @@
+
+## Install the Astronomer CLI
+
+Astronomer's [open source CLI](https://github.com/astronomer/astro-cli) is the easiest way to run Airflow on your local machine. You'll need the CLI to complete this quickstart and create a local Airflow instance with a dedicated Webserver, Scheduler, and Postgres database.
```
```suggestion
Astronomer's [open source CLI](https://github.com/astronomer/astro-cli) is the easiest way to run Airflow on your local machine. You'll need the Astronomer CLI to create a local Airflow instance with a dedicated Webserver, Scheduler, and Postgres database.
```

paolaperaza commented on ac/next/01_quickstart.md:
````diff
@@ -0,0 +1,166 @@
+If the installation was successful, the version of the CLI that you installed will appear in the output:
+
+```
+Astro CLI Version: 0.15.0
````
```suggestion
Astro CLI Version: 0.23.3
```

paolaperaza commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,166 @@
+
+## Install Astronomer Core via Docker
+
+To install and run the Astronomer Core distribution:
```
```suggestion
To install and run the Astronomer Core distribution, follow the steps below.
```

jwitz replied:
Reworded slightly

paolaperaza commented on ac/next/01_quickstart.md:
````diff
@@ -0,0 +1,166 @@
+    └── requirements.txt # For any Python packages
+    ```         
+
+3. Start Airflow on your local machine by running the following command:
````
```suggestion
3. Start Airflow locally by running the following command:
```

paolaperaza commented on ac/next/01_quickstart.md:
````diff
@@ -0,0 +1,166 @@
+4. Verify that the 3 Docker containers were succesfully created by running:
+
+    ```sh
+    docker ps
````
```suggestion
   $ docker ps
```

paolaperaza commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,166 @@
+
+## Install via Python Wheel
+
+We also distribute Astronomer Core as a python wheel available on PyPi. This distribution can be run locally via the same mechanisms used to run the Apache Airflow package.
```
```suggestion
In addition to the Docker-based distribution, Astronomer Core is built as a Python Wheel available on PyPi. This distribution can be run locally via the same mechanisms used to run the Apache Airflow package.
```

paolaperaza commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,166 @@
+
+We also distribute Astronomer Core as a python wheel available on PyPi. This distribution can be run locally via the same mechanisms used to run the Apache Airflow package.
+
+1. Run `export AIRFLOW_HOME=~/airflow` to give Astronomer Core a home root directory on your machine.
```
```suggestion
1. Run `$ export AIRFLOW_HOME=~/airflow` to give Astronomer Core a home root directory on your machine.
```

jwitz replied:
Assuming these suggestions are nulled given our recent Slack convo

paolaperaza commented on ac/next/01_quickstart.md:
````diff
@@ -0,0 +1,166 @@
+2. Run the following command to get the latest version of the distribution on your machine:
+
+    ```sh
+    PIP_EXTRA_INDEX_URL='https://pip.astronomer.io/simple' pip3 install -U 'astronomer-certified==1.10.10.*'
````
Does this assume the latest `Airflow 1.10.x`? Should we be prescriptive here in proposing 2.0, or at least call this out?

paolaperaza replied:
```suggestion
   $ PIP_EXTRA_INDEX_URL='https://pip.astronomer.io/simple' pip3 install -U 'astronomer-certified==1.10.10.*'
```

paolaperaza commented on ac/next/01_quickstart.md:
````diff
@@ -0,0 +1,166 @@
+    PIP_EXTRA_INDEX_URL='https://pip.astronomer.io/simple' pip3 install -U 'astronomer-certified==1.10.10.*'
+    ```
+
+    Note that if you'd like to install a specific version, you can do so by appending the version tag via the following syntax:
````
```suggestion
    To install a specific version of Astronomer Core, append the version tag via the following syntax:
```

paolaperaza commented on ac/next/01_quickstart.md:
````diff
@@ -0,0 +1,166 @@
+    Note that if you'd like to install a specific version, you can do so by appending the version tag via the following syntax:
+
+    ```sh
+    PIP_EXTRA_INDEX_URL='https://pip.astronomer.io/simple' pip3 install -U 'astronomer-certified==1.10.10-1'
````
```suggestion
   $ PIP_EXTRA_INDEX_URL='https://pip.astronomer.io/simple' pip3 install -U 'astronomer-certified==1.10.10-1'
```

paolaperaza replied:
Should this be `astronomer-certified==<airflow-version>`?

paolaperaza commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,166 @@
+
+3. Run `$ airflow initdb` to build out your project directory and initialize a lightweight SQLite database for Airflow.
+
+4. Run `$ airflow webserver -p 8080` to spin up the Webserver on your `localhost:8080`.
```
```suggestion
4. Run `$ airflow webserver -p 8080` to provision the Webserver on your `localhost:8080`.
```

paolaperaza commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,166 @@
+
+4. Run `$ airflow webserver -p 8080` to spin up the Webserver on your `localhost:8080`.
+
+5. Run `$ airflow scheduler` to spin up the Scheduler.
```
```suggestion
5. Run `$ airflow scheduler` to provision the Scheduler.
```

paolaperaza commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,166 @@
+
+5. Run `$ airflow scheduler` to spin up the Scheduler.
+
+## Next Steps
```
This guide seems to stop at the point where you have the Astronomer CLI installed, but it feels a little bit like we're leaving the user hanging. We need to figure out a connection here with the "Production" story", but I'd consider adding an extra section for:

- "Apply Changes to your Airflow Project" - https://www.astronomer.io/docs/cloud/stable/develop/cli-quickstart#apply-changes-to-your-airflow-project

Once we have CLI reference guide, will be critical to link here too.

paolaperaza reviewed (changes requested):
@jwitz Finally got a chance to take a deeper look at this. Let's focus on changes to Local Quickstart for now and merge that separately to the changes in "Production Core".

Some small comments on typos and copy suggestions, but a few comments as well on some additional content and structure we may be able to add. Let me know what you think.

jwitz commented on ac/next/01_quickstart.md:
```diff
@@ -0,0 +1,166 @@
+---
+title: "Astronomer Core Quickstart"
```
I think `local` is implied by the nature of a quickstart. Also would that not imply the existence of a production quickstart, which we don't really have?

paolaperaza reviewed (commented):
Some final comments here @jwitz - let me know what you think so I can give a final approval and get this thing merged as an MVP

jwitz commented on ac/next/01_quickstart.md:
````diff
@@ -0,0 +1,170 @@
+2. Install Astronomer Certified using pip3. For example, to install Astronomer Certified for Airflow 2.0.0, you would run:
+
+    ```sh
+    PIP_EXTRA_INDEX_URL='https://pip.astronomer.io/simple' pip3 install -U 'astronomer-certified==2.0.0.*'
````
It's more than the Airflow version through, right? I guess I don't understand the function of the asterisk at the end of the version number.

kaxil replied:
Asterisk basically tells give me the latest version available in 2.0.0 AC series, so this should match 2.0.0-1, 2.0.0-2, 2.0.0-3 etc

jwitz commented:
LGTM @paolaperaza, just had one question about the pip install command but otherwise I think this is good to merge!
